### ISSUE TITLE
Replacing workspaces with windows native Virtual Desktops - POLL

### ISSUE THREAD

**Issue by N1x0** [workspacer/workspacer#219]
I'm currently checking the viability of replacing the current workspaces with the windows-native VirtualDesktops. 

Reason for this is that there are currently some issues with UWP apps (e.g. Power Bi) shutting down when leaving the workspace they are on, (see #97) and also with moving electron/chromium windows turning black (#171). This does not happen with the virtual desktops from my testing.

I found some viable alternatives for readily-available c# API's that allow access and found [this](https://github.com/MScholtes/VirtualDesktop). However compatibility is limited to windows 10.

Before going further and creating an alternative branch with VirtualDesktops replacing the current workspace implementation I first wanted to see how many users' installation would be affevted by this change.

In general I am quite optimistic on the development of virtual desktops, as new features have been showcased in the windows 11 release announcement. I think it'd be better for workspacer to use a solution provided by the OS and integrating it rather than duplicating the functionality.

Feedback and comments are most welcome!

[![](https://api.gh-polls.com/poll/01F98TZYRHBWPQ1JF2FTJHF919/Windows%2010)](https://api.gh-polls.com/poll/01F98TZYRHBWPQ1JF2FTJHF919/Windows%2010/vote)
[![](https://api.gh-polls.com/poll/01F98TZYRHBWPQ1JF2FTJHF919/Windows%208.1)](https://api.gh-polls.com/poll/01F98TZYRHBWPQ1JF2FTJHF919/Windows%208.1/vote)
[![](https://api.gh-polls.com/poll/01F98TZYRHBWPQ1JF2FTJHF919/Windows%208.0)](https://api.gh-polls.com/poll/01F98TZYRHBWPQ1JF2FTJHF919/Windows%208.0/vote)
[![](https://api.gh-polls.com/poll/01F98TZYRHBWPQ1JF2FTJHF919/Windows%207)](https://api.gh-polls.com/poll/01F98TZYRHBWPQ1JF2FTJHF919/Windows> %207/vote)

**Comment by sitiom** [comment]
This was first mentioned in #62:
> * **What about use of native virtual desktops?**: does it seems reasonable? I think that both virtual desktops and tagged space could coexist. Not sure if viable or really open some opportunity, but it seems that relying on a native feature could give some plus: Win+TAB giving a snapshots of "tagged" windows for example...


>  However compatibility is limited to windows 10.

Well according to the workspacer homepage, it is "a tiling window manager for Windows 10", isn't it? Builds right now are built on the `win10-x64` platform as well.

Anyways, +1 for this. I have used [win3wm](https://github.com/McYoloSwagHam/win3wm) which makes use of virtual desktops and haven't encountered the issues mentioned above. This would also ultimately fix #37 as you are just switching between virtual desktops instead of hiding them as well.

**Comment by dalyIsaac** [comment]
For chromium turning black, this has been triggered for me when changing layout engines - it may be related to moving windows. It'd be interesting to see what PowerToys does for laying out windows vs. what workspacer currently does.

For virtual desktops, do you know if each monitor can have a separate virtual desktop in Windows 11? Currently, in Windows 10 switching virtual desktops switches for all monitors, which would mean monitors could no longer have independent workspaces open.

**Comment by josteink** [comment]
> I found some viable alternatives for readily-available c# API's that allow access and found this. However compatibility is limited to windows 10.

Workspacer is *officially* Windows 10+ only, so that’s not an issue.

**Comment by N1x0** [comment]
> 
> 
> For chromium turning black, this has been triggered for me when changing layout engines - it may be related to moving windows. It'd be interesting to see what PowerToys does for laying out windows vs. what workspacer currently does.
> 
> For virtual desktops, do you know if each monitor can have a separate virtual desktop in Windows 11? Currently, in Windows 10 switching virtual desktops switches for all monitors, which would mean monitors could no longer have independent workspaces open.

I have checked the PowerToys repo and found [this similar issue](https://github.com/microsoft/PowerToys/issues/3536). Supposedly it was fixed but it seems like it's not quite gone.

I forgot about the issues with multiple displays on win10, but there is some talk of improved [multi-monitor support](https://www.gsmarena.com/windows_11_will_finally_introduce_advanced_multimonitor_settings-news-49681.php), but it's only slightly related and not a real fix. Sadly this makes this unworkable for multiple displays as it seems.

**Comment by imswebra** [comment]
I know [Amethyst Windows](https://github.com/glsorre/amethystwindows) uses the native virtual desktops and has multi-monitor support, but I have not tested multi-monitor support myself. It seems to be a very early project, but perhaps there is something to learn from it. A couple nice things I noticed when using it were that electron apps behaved more regularly and the desktop switched automatically when doing something like opening an link (automatically switched to the desktop with my browser).

**Comment by JakobBruenker** [comment]
@N1x0 I don't have a good source for this, but [this video](https://youtu.be/odZSCdNTFPw?t=489) at 8:10 says "On a per-display basis, you can have multiple virtual desktops that you can switch between" in Windows 11, which seems to be exactly what we need and not covered by the link you posted.

(Also, by the way, I think I may have accidentally voted for Win 8.1 since I didn't realize clicking on it counts a vote instead of taking you to a page to vote)

**Comment by N1x0** [comment]
> 
> 
> I know [Amethyst Windows](https://github.com/glsorre/amethystwindows) uses the native virtual desktops and has multi-monitor support, but I have not tested multi-monitor support myself. It seems to be a very early project, but perhaps there is something to learn from it. A couple nice things I noticed when using it were that electron apps behaved more regularly and the desktop switched automatically when doing something like opening an link (automatically switched to the desktop with my browser).

Thanks for mentioning Amethyst Windows, I had a look to see what the multi-monitor support includes and sadly it doesn't solve our issues with workspacer as it also doesn't fix the inability to move virtual desktops per monitor. Also I very quickly experienced the bug we have with chromium windows turning black (see screenshot)
![image](https://user-images.githubusercontent.com/47623143/123780440-6489a580-d8fd-11eb-9fd7-23afe0ec054b.png)


So I don't think Amethyst has the solutions to our problem, which is very unfortunate because the author based their work on the VirtualDesktopManager repo's I had looked at as well.

@JakobBruenker 

I hope that that's indeed the case, it's been an issue for over half a decade so would be good of MS to get around to fixing it. 
Once that has happened I think it should be possible fully replace the in-built workspaces with the native VirtualDesktops. Whether that will be back-ported to Win10 remains to be seen.


I have also looked at [win3wm ](https://github.com/McYoloSwagHam/win3wm) but no luck there either, it supports different bars but switching between virtual desktops again moves them on both monitors

**Comment by dalyIsaac** [comment]
I don't know where Linus Tech Tips got that tidbit from - it hasn't been in any of the recent Windows 11 coverage.

However, I did find a report from [Winaero](https://winaero.com/windows-10-will-soon-have-individual-virtual-desktops-per-display/) which indicates that it is a feature which Microsoft has added to internal builds. Nevertheless, it's not currently in Windows 11 (as of 22000.51).

**Comment by lauverityhill** [comment]
It seems the best course of action is to wait for Microsoft's Task View API to allow multiple Virtual Desktops on each monitor. Switching to Virtual Desktops right now would make multi-monitors quite irritating to use. Though starting work now to future proof Workspacer isn't a bad idea (on a separate branch I guess), since I believe Virtual Desktops would be the long term solution anyway.

**Comment by N1x0** [comment]
> It seems the best course of action is to wait for Microsoft's Task View API to allow multiple Virtual Desktops on each monitor. Switching to Virtual Desktops right now would make multi-monitors quite irritating to use. Though starting work now to future proof Workspacer isn't a bad idea (on a separate branch I guess), since I believe Virtual Desktops would be the long term solution anyway.

I agree with this approach would be nice to have a virtualdesktop fork already set-up for when the changes come through and those using single monitors could already take advantage of this now.

**Comment by riotrah** [comment]
I offer the following points, as a fan of TWMs and of this incredible repo, and as someone who is frustrated with the lack of usable TWMs for Win 10/11, and someone whose programming languages of choice are not popular among TWM repos:

1. I agree as well that native integration is more future-proof, as has already been argued in this thread.
   a. Certainly with strong competition from MacOS and nix UX, and as evidenced by PowerToys FancyZones, MS is invested in advancing window management and vDesktops in general
2. I think native integration is _at _least_ worth implementing as a toggleable, caveat-warned alternate even outside of future-proofing, because:
3. **Not necessarily everyone wants separate desktops per monitor**
   a. I myself do not
   a. Despite seeing people argue as much in forums, issues, etc, it's not clear that the _majority_ of virtual desktop users are indeed per-monitor desktop advocates.
   b. MacOS offers the **option** of shared vs per-monitor, with clearly announced caveats ( eg a single menu bar or one per monitor? )
4. Nor will everyone who suffers from the above listed bugs due to non-native workspaces ( electron black windows etc ) be _unwilling_ to give up separate workspaces, in exchange for a bug-free/reduced experience.
   a. I could imagine users that get more use out of frequent electron app (which is very popular) usage, versus use of workspacer's per-monitor workspace paradigm, might be okay with the compromise, esp if its caveats are clear and acknowledged, or is optional
5. Less important but not trivial: Win 10/11's swipe gestures for virtual desktops are pretty delightful and W11's new Win+Tab view is quite pretty, and so native integration _potentially_ elevates the sleekness of any TWM that makes use of them.

I hope this wasn't too verbose to the point of noisiness, please lmk and I'll edit. I want to express gratitude again for this awesome, domain-leading work, and proactive, open-minded philosophy that the org is taking. 

As I set up my own config, hopefully I'll pick up some non-trivial C# skills and also contribute in the near future!
